### ISSUE TITLE
memory.swappiness may not be writeable on some kernels

### ISSUE THREAD

**Issue by estesp** [docker/libcontainer#639]
PR #622 adding the capability for writing to the `memory.swappiness` setting within a cgroup.  Because of the check (>= 0 and <= 100), even if the cgroup object has no intended setting, by default, I believe it will be 0 (which is a Go default for `int64`, and, if using from Docker, there is no initial setting in docker's "execdriver" template), and therefore, every container start will try and write to the cgroup sysfs entry.

This has two problems: 1) I'm going to assume that the intent of adding the behavior was not to always set swappiness to zero on every container/cgroup (which is I think what happens right now), and 2) setting memory.swappiness in hierarchical cgroups was disabled (you get EINVAL) until this patch: torvalds/linux@3dae7fec5e884a4e72e5416db0894de66f586201 which appears to be in 3.16 and newer kernels.

On Ubuntu 14.04.02 LTS for example, using Docker with an updated libcontainer, even without the "UI" elements for interacting with `memory.swappiness`, every container start fails because of the cgroup write EINVAL result from trying to write zero to `memory.swappiness` on that older kernel.

A fix for problem 1 would be to somehow default an "unset" `MemorySwappiness` to `-1` so writes aren't inadvertently done every time with a zero value.  I think we need to discuss problem 2 as far as either graceful fallback on failure with EINVAL or removing the feature until it is available in the lowest supported kernel.

**Comment by rjnagal** [comment]
-1 for 1) sounds good.
For the second part, I think libcontainer should return an error if a user asks for swappiness when its not available. Docker daemon can ignore requests when swappiness is not enabled, just the way it does for mem swap.

**Comment by dqminh** [comment]
WDYT about changing the default type for cgroups config value to `*int64`. It's a bit more troublesome to consume, but it allows us to properly detect whether values are set by user.

**Comment by dqminh** [comment]
For 2), I'm not sure how we can properly detect this for older kernel. If 1) is fixed, then i think that failing to set user-defined swappiness value is a fatal error and the container should not be started (i.e., there's no approriate fallback)

**Comment by ktraghavendra** [comment]
@estesp @dqminh @cpuguy83 
It is correct that we do not want to set the mem swappiness value to zero for everybody.

Assuming that failing to set the swappiness value when user has asked to do so..
we have
case A: explicit setting of memory swappiness value on a supported kernel
Container does not start
case B: explicit setting of memory swappiness on a  unsupported kernel
Container does not start

case C: Implicit setting of memory swappiness value on a supported kernel
This should not be done. 

case D: Implicit setting of memory swappiness value on a unsupported kernel.
 once case C is fixed D is not a problem

How to handle (D and container starts normally) if (C) is not fixed.
We could check (if root memory swappiness exists and root use_hierarchy is set) then
 failure in trying to set implicit value should be allowed.

@estesp with https://github.com/docker/docker/pull/14030, do we avoid setting value of swappiness and fix (C) ?

**Comment by estesp** [comment]
@ktraghavendra yes, the Docker PR means there will be no inadvertent write to swappiness until a PR is merged which allows the user to specify swappiness (your "UI PR" as I called it).  Once that PR is merged, then the only failures on the Docker side will be when a user explicitly tries to set a swappiness value, but the kernel is old enough to predate the loosened restriction on cgroup hierarchy use + swappiness writes.

That case (failure due to older kernel) we could handle only on the Docker side (similar to what we do when OOM and memory restrictions don't work because those features aren't turned on via kernel command line, etc.

The only case that remains that isn't solved is regarding a "new" libcontainer consumer (outside of nsinit and docker).  Using the "out of the box" implementation for configs, that consumer of libcontainer will get automatic writes attempted with `swappiness = 0` unless they have a "template init" style that sets up a better default value that matches the 2 PRs I've submitted for docker and nsinit.

I believe this is where @dqminh is suggested possibly moving to an `*int64` as it will require explicit initialization to determine that a write is requested.

**Comment by ktraghavendra** [comment]
@estesp @dqminh , Thanks .. yes moving to *int64 is a good option. So in libcontainer we do not return error if point is still a null. Will work on that patch. From that we do not have to even check the older kernel failure stuff.
